### PR TITLE
Fix getByKey to return all errors

### LINKED ISSUE #15103: imagestream_controller uses an unusual construct for getByKey
https://github.com/openshift/origin/pull/15099

Bug was due to returning nil, nil and not handling properly.  We generally never do that, should be refactored to be consistent with other controllers.

### PR DESCRIPTION
**Pull request by soltysh**
Fixes #15103.

@smarterclayton || @deads2k ptal

### CODE REVIEW

File: pkg/deploy/controller/deployer/factory.go [hunk 1]
```diff
@@ -217,6 +217,7 @@ func (c *DeploymentController) getByKey(key string) (*v1.ReplicationController,
 	}
 	rc, err := c.rcLister.ReplicationControllers(namespace).Get(name)
 	if errors.IsNotFound(err) {
+		// TODO tnozicka: this is not normal and should be refactored
```
Comment (soltysh) [inline pkg/deploy/controller/deployer/factory.go]: I've created https://github.com/openshift/origin/issues/15928 for you to address this.

Comment (deads2k) [inline pkg/deploy/controller/deployer/factory.go]: What is the value in having this function now?  I expected to see something like this: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/replicaset/replica_set.go#L553-L561

Comment (mfojtik) [inline pkg/deploy/controller/deployer/factory.go]: @tnozicka FYI

**Comment by soltysh** [comment]
Rebased. @smarterclayton ping

File: pkg/image/controller/imagestream_controller.go [hunk 2]
```diff
@@ -118,14 +118,15 @@ func (c *ImageStreamController) processNextWorkItem() bool {
 
 	stream, err := c.getByKey(key.(string))
```
Comment (deads2k) [inline pkg/image/controller/imagestream_controller.go]: Seems like this function should just be removed.

**Comment by soltysh** [comment]
@deads2k comments addressed, ptal

File: pkg/image/controller/imagestream_controller.go [hunk 3]
```diff
@@ -116,14 +116,21 @@ func (c *ImageStreamController) processNextWorkItem() bool {
 	}
 	defer c.queue.Done(key)
 
-	stream, err := c.getByKey(key.(string))
+	namespace, name, err := cache.SplitMetaNamespaceKey(key.(string))
 	if err != nil {
-		utilruntime.HandleError(fmt.Errorf("Error syncing image stream: %v", err))
+		glog.V(2).Infof("Unable to split namespace key for key %q: %v", key, err)
```
Comment (deads2k) [inline pkg/image/controller/imagestream_controller.go]: `utilruntime.HandleError`, then forget and drop the key.  The key is never going to get fixed.

File: pkg/image/controller/imagestream_controller.go [hunk 4]
```diff
@@ -116,14 +116,21 @@ func (c *ImageStreamController) processNextWorkItem() bool {
...
 		c.queue.AddRateLimited(key)
 		return true
 	}
-	if stream == nil {
-		c.queue.Forget(key)
+	stream, err := c.lister.ImageStreams(namespace).Get(name)
```
Comment (deads2k) [inline pkg/image/controller/imagestream_controller.go]: Can we reshape this like upstream, where the sync function takes the key and does the work there?  it keeps the error handling code a little cleaner and more standard.

**Pull request merged**